pyanaconda/modules/subscription/subscription.py

```diff
@@ -0,0 +1,170 @@
+from pyanaconda.core.signal import Signal
+
+from pyanaconda.modules.common.base import KickstartService
+from pyanaconda.modules.common.structures.subscription import SystemPurposeData
+
+from pyanaconda.modules.subscription import system_purpose
+from pyanaconda.modules.subscription.kickstart import SubscriptionKickstartSpecification
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class SubscriptionService(KickstartService):
+    """The Subscription service."""
+
+    def __init__(self):
+        super().__init__()
+
+        # system purpose
+
+        self._valid_roles = []
+        self._valid_slas = []
+        self._valid_usage_types = []
+
+        self._system_purpose_data = SystemPurposeData()
+        self.system_purpose_data_changed = Signal()
+
+        self._load_valid_system_purpose_values()
+
+        # FIXME: handle rhsm.service startup in a safe manner
+
+    @property
+    def kickstart_specification(self):
+        """Return the kickstart specification."""
+        return SubscriptionKickstartSpecification
+
+    def process_kickstart(self, data):
+        """Process the kickstart data."""
+        log.debug("Processing kickstart data...")
+
+        # system purpose
+        #
+        # Try if any of the values in kickstart match a valid field.
+        # If it does, write the valid field value instead of the value from kickstart.
+        #
+        # This way a value in kickstart that has a different case and/or trailing white space
+        # can still be used to preselect a value in a UI instead of being marked as a custom
+        # user specified value.
+        role = system_purpose.process_field(data.syspurpose.role, self.valid_roles, "role")
+        sla = system_purpose.process_field(data.syspurpose.sla, self.valid_slas, "sla")
+        usage = system_purpose.process_field(data.syspurpose.usage,
+                                             self.valid_usage_types,
+                                             "usage")
+
+        system_purpose_data = SystemPurposeData()
+        system_purpose_data.role = role
+        system_purpose_data.sla = sla
+        system_purpose_data.usage = usage
+
+        if data.syspurpose.addons:
+            # As we do not have a list of valid addons available, we just use what was provided
+            # by the user in kickstart verbatim.
+            system_purpose_data.addons = data.syspurpose.addons
+
+        self.set_system_purpose_data(system_purpose_data)
+
+    def setup_kickstart(self, data):
+        """Return the kickstart string."""
+        log.debug("Generating kickstart data...")
+
+        # system purpose
+        data.syspurpose.role = self.system_purpose_data.role
+        data.syspurpose.sla = self.system_purpose_data.sla
+        data.syspurpose.usage = self.system_purpose_data.usage
+        data.syspurpose.addons = self.system_purpose_data.addons
+
+        return str(data)
```

Comment:
The method `setup_kickstart` modifies the data object. It doesn't return anything.

Reply:
OK, dropping that.